designs/function-node-lifecycle/README.md

```diff
@@ -53,6 +53,50 @@ Function node declares its required NPM module list to runtime by using new API.
 
 Because current Node-RED implementation starts flow execution before message handler callback is ready, we may need to modify runtime execution model in order to make function nodes wait for receiving messages before required NPM modules become ready to use.
 
+#### Additional settings for external npm module in `settings.js`
+
+Add a top-level property named `externalModule` to `settings.js`.  It points to a object with following properties.
+
+| name | type   | required | description  |
+| :--- | :----- | :------- | :----------- |
+| mode | string | yes      | one of `"auto"`, `"auto-update"`, `"manual"`, or `"none"` (default: `"manual"`)<br/>- `auto`: automatically install requested NPM module if not installed.<br/>- `auto-update`: same as `auto` but with automatic update if newer version is available.<br/>- `manual`: users can manually install requested NPM module from user settings panel.<br/>- `none`: installation of external NPM modules is not allowd |
+| whiteList | array of string | no | List of regular expressions that matches module's NPM registry specification. External libraries that matches this list **can** be used in `Function` nodes. |
+| blackList | array of string | no | List of regular expressions that matches module's NPM registry specification. External libraries that matches this list **can not** be used in `Function` nodes. |
+
+`whiteList` precedes over `blackList`.
+
+```
+    // Example of settings.js
+    externalModules: {
+        mode: "manual",
+        whiteList: [
+            "^fs-ext$",       // allow fs-ext
+            "^qrcode@1.4.4$"  // allow qrcode version 1.4.4
+        ],
+        blackList: [
+            ".*"              // disallow others 
+        ]
+    },
+```
+If the `externalModule` property not exists, `none` mode and blackList of `".*"`, meaning not allow installation and use of external module, is expected as a default.
+
+#### New API for requesting NPM module
+
+- `RED.nodes.registerRequiredModules`(*module*)
+
+Requests Node-RED runtime to install NPM module *module*.  Returns a promise that resolves to required module object or rejects if installation failed.  NPM module installation is performed according to specification in `externalModule` property of `settiongs.js` described above.
+
+If `auto` mode is specified, execution of a function node setup or body that require NPM modules waits for completion of their installation.
+
```

Comment:
Current implementation of the Function node retains the messages accepted during the execution of initialization code. As you suggested, it looks good to wait for the installation to complete in the runtime. 
Because I want the NPM module to have additional information, I made the value of `dynamicModuleList` to also have a function value.